airflow/jobs.py

```diff
@@ -990,11 +989,14 @@ def _execute_task_instances(self,
         # Get all the queued task instances from associated with scheduled
         # DagRuns.
         TI = models.TaskInstance
+        DR = models.DagRun
         task_instances_to_examine = (
             session
             .query(TI)
             .filter(TI.dag_id.in_(simple_dag_bag.dag_ids))
             .filter(TI.state.in_(states))
+            .outerjoin(DR, and_(DR.dag_id==TI.dag_id, DR.execution_date==TI.execution_date))
```

Comment:
@AllisonWang  Here it is much faster. The join is super fast because there are good indices for it. The old logic is pretty slow because it makes a separate query for each task instance, and it is done synchronously. The amount of time taken by the join is likely <100ms, whereas the old approach is (database roundtrip + query_time) * (number of TIs). In practice, probably a minute in some of the worst cases.